### DISCUSSION TITLE
API at runtime

### DISCUSSION THREAD

**Discussion by donald-richardson-dxd** [reduxjs/redux-toolkit#3086] [Q&A]
Is there a way to create an API dynamically at runtime from the GraphQL schema? Not using code generation but rather perhaps through generics? We're creating a CMS that allows users to create their own content types through no code. It would be nice if they could then access those models via graphql without having to generate code files. It would just automagically work somehow.

This would be using Typescript. Any pointers would be helpful.

**Comment by markerikson** [comment]
I'm not really sure what you're asking here.

TypeScript "Generics" are a compile-time thing, not runtime behavior.

It's certainly possible to generate endpoint descriptions via code and call `injectEndpoints`, which is what the codegen packages do in the first place.

That said, "just automagically work somehow" really isn't a clear description of what you're really trying to do :)

**Comment by donald-richardson-dxd** [comment]
Fair enough. And that vagueness is a side-effect of my ignorance. I'm just trying to figure out how its possible to set up RTK Query without knowing the data models ahead of time. So I wouldn't know the type Pokemon at compile time. If I injectEndpoints after that is it possible to have the state syncing functionality without knowing any of the types ahead of time?

**Reply by markerikson** [reply]
Yes, in that the endpoints would have been added, the _runtime_ behavior would exist, and the hooks would have been generated.  But none of the TS types would exist.

How are you expecting someone to actually make use of this functionality?

**Reply by donald-richardson-dxd** [reply]
Good question. I'm trying to create an SDK and I just don't want the end-user to have to set up a complicated type definition when all of the necessary information already exists in the schema. I know that codegen solves this but it's also unwieldy, in my particular case. It would be great if the end user could just do something like the following:

```
class MyModel extends GenericModel {
  constructor(modelName: string) {
    super(modelName)
  }

  getPrettyName() {
     const data = this.getData()
     return `${data["firstName"] data["lastName"]}`
  }
}
const myModel: MyModel = sdk.createModel("myModelName")
let {data, status, error} = myModel.getAll()
let {data, status, error} = myModel.getById(id)
let prettyName = myModel.getPrettyName()
```

**Reply by markerikson** [reply]
Honestly, that doesn't look anything like how React or Redux typically get used.

I'm really not sure RTK Query is the right tool for what you seem to have in mind.

**Reply by donald-richardson-dxd** [reply]
Perhaps. Unfortunately, neither does any other tool offhand. And I really don't want to reinvent the wheel when it comes to data synchronization. I'm amenable to it looking different, I just don't know the types ahead of time. But I do have enough information at runtime (by retrieving the GraphQL schema) to know the data shape.
